### PR TITLE
Added Application Release History Logs

### PR DESCRIPTION
**Pull request by pxwxnvermx**
## Product Description

This feature adds logs for release related activity on applications and the logs are displayed on the Application Release Logs tab. Logs are added for all releases that are made along with the status changes from In-Test to Released.

Screenshot 2022-11-14 at 10 23 30 AM

## Technical Summary
[Jira Ticket](https://dimagi-dev.atlassian.net/browse/INDIV-67)

## Feature Flag
APPLICATION_RELEASE_LOGS

## Safety Assurance
Tested locally

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:72]: We would need to add translation for these strings?

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:73]: i think we would need to render time based on user timezone. There is a timezone set at user level and also at project level. Does this do that for the user's local timezone in Javascript itself?

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:85]: What is this function for? A better name would be useful to highlight its purpose

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py:289]: nit: Can possibly extract this out in a function.

Comment (mkangia) [reply corehq/apps/app_manager/views/releases.py:289]: Also, can we add a test for this?

**Review by mkangia — COMMENTED** [review]
Smooth! left some feedback.

File: corehq/apps/app_manager/models.py
Comment (orangejenny) [inline corehq/apps/app_manager/models.py:6252]: Curious why this is being stored, instead of being looked up when the logs are pulled. Generally we shouldn't be storing denormalized data. (Domain is arguably redundant, but it's so useful for sql dump and domain deletion.)

**Review by orangejenny — COMMENTED** [review]
Nice! PR description & screenshots would be nice when you're ready (recognizing that this is still in draft state).

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (pxwxnvermx) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js]: This function is used for getting the endpoint associated with a slug. Renamed it to getApiEndpoint.

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (pxwxnvermx) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js]: Yup, fixed this with time based on user and project timezone on the data returned itself. Removed JS logic as there is no need to change the timestamp string in JS now.

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js]: Why does this function exist? Can't you call `hqImport("hqwebapp/js/initial_page_data").reverse("paginate_release_logs")` where you're currently calling `self.getApiEndpoint("paginate_release_logs")`?

File: corehq/apps/app_manager/models.py
Comment (mkangia) [reply corehq/apps/app_manager/models.py:6252]: > Generally we shouldn't be storing denormalized data

@orangejenny Can you expand on this more? 

Moreover, I think the intent here was to avoid the lookup though we can use `cached_user_id_to_user_display` as well.

Comment (orangejenny) [reply corehq/apps/app_manager/models.py:6252]: I mean that we shouldn't store data that con be looked up based on other data, since it can get out of sync. (Admittedly, in this case I'm not 100% certain a user can change their email address, since we use it for username, so I mean this as a general best practice.)

I agree about using `cached_user_id_to_user_display`, relying on code solutions like caches is usually better than storing more data.

Comment (pxwxnvermx) [reply corehq/apps/app_manager/models.py:6252]: @orangejenny @mkangia thanks for the input, i will make the changes.

File: corehq/apps/app_manager/models.py
Comment (mkangia) [inline corehq/apps/app_manager/models.py:6252]: Should we name this `created_at` or something similar so its clear we are talking about the creation timestamp and not say, update?

File: corehq/apps/app_manager/models.py
Comment (mkangia) [inline corehq/apps/app_manager/models.py:6253]: can we make this `action` instead? I am thinking for now we just have "Released" and "In-Test" but if we want to track any other action in future, we would need to modify the model, like say for "delete"? 
This makes me think if we should track delete as well?

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:72]: Interesting. What does `ko.mapping.fromJS` do?

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:85]: what arguments is this working on? cant you directly use the registered url in view `paginate_release_logs`?

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:98]: nit: unexpected blank line.

File: corehq/apps/app_manager/templates/app_manager/partials/releases/app_release_logs.html
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/partials/releases/app_release_logs.html:43]: nit: I think you would need to put a blank line here. I believe that is already by editor configs we have setup for use
https://github.com/dimagi/commcare-hq/blob/9cacc528b3d098d2ef0c1bedd76dcff0da3bbe2f/.editorconfig#L7

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py:26]: nit: Did you already run isort on this?

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py:676]: nit: this could be a constant defined in `app_manager/const`

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py:700]: can this be rendered together in `ui_string` in a single call? Not sure why we need to do date and time separately.

File: corehq/toggles/__init__.py
Comment (mkangia) [inline corehq/toggles/__init__.py:2380]: nit: `with user and version history` seems redundant here.

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [reply corehq/apps/app_manager/views/releases.py:700]: We already spoke about this offline,so just noting to remember for code review. 
We should not modify the original object.

**Review by mkangia — COMMENTED** [review]
Looking good! Just left some feedback.

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (pxwxnvermx) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:72]: `ko.mapping.fromJS` generated knockout viewmodels from javascript objects. That makes it easier, and eliminates the need to make the viewmodels ourselves.

File: corehq/apps/app_manager/models.py
Comment (pxwxnvermx) [reply corehq/apps/app_manager/models.py:6252]: Yup, will change `timestamp` to `created_at` as it basically reflects the timestamp of creation of the log and also the time at which action was performed.

File: corehq/apps/app_manager/models.py
Comment (pxwxnvermx) [reply corehq/apps/app_manager/models.py:6253]: Yeah, this makes sense, better to have an `action` field than `is_released`, that will make it flexible to track other actions on the app releases.

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (pxwxnvermx) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:85]: this function takes the registered url slug i.e `paginate_release_logs` and calls the reverse function in initial_page_data which gives back the url of the endpoint.
https://github.com/dimagi/commcare-hq/blob/485a41cb902eae8acc6e5ffdf477104f49b9106c/corehq/apps/hqwebapp/static/hqwebapp/js/initial_page_data.js#L88

Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:85]: Are there any arguments that are actually passed? It looks like the `page` and other params are passed as a part of the data object in the request and not in the url. Can you confirm what arguments are passed here?

Comment (pxwxnvermx) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js:85]: `getApiEndpoint` is a variadic function, look at the loop that is parsing the variable passed as args, i picked this function up from another file. @orangejenny suggested in a review comment above to directly use the `reverse` function which is wrapped by this function. That looks like a better way to go than having a function for a getting a single api endpoint.

File: corehq/apps/app_manager/models.py
Comment (pxwxnvermx) [reply corehq/apps/app_manager/models.py:6273]: ok, that function isn't used anywhere other from this, will change this to use ACTION_DISPLAY directy and remove that function

File: corehq/apps/app_manager/templates/app_manager/partials/releases/app_release_logs.html
Comment (pxwxnvermx) [reply corehq/apps/app_manager/templates/app_manager/partials/releases/app_release_logs.html:24]: this value is generated from `created_at` timestamp that is on the model, and is present as `created_at_string` property on the data that is returned to the frontend.

Comment (mkangia) [reply corehq/apps/app_manager/templates/app_manager/partials/releases/app_release_logs.html:24]: Cool. My understanding is that the problem is that we already have "created_at" on the model that conflicts.
Sounds good to retain a separate variable for display

File: corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js
Comment (mkangia) [reply corehq/apps/app_manager/static/app_manager/js/releases/app_view_release_manager.js]: I thought it was a mistype! Thanks @pxwxnvermx !

File: corehq/apps/app_manager/tests/views/test_release_build.py
Comment (mkangia) [inline corehq/apps/app_manager/tests/views/test_release_build.py:44]: does this include app and app build delete as well? I am not sure if we delete apps separately in tests as well.

File: corehq/apps/app_manager/tests/views/test_release_build.py
Comment (mkangia) [inline corehq/apps/app_manager/tests/views/test_release_build.py:47]: nice that we added a test for the default behavior as well :100:

File: corehq/apps/app_manager/tests/views/test_release_build.py
Comment (mkangia) [inline corehq/apps/app_manager/tests/views/test_release_build.py:65]: can we check for user_id as well? We already covered app_id in the filter so that was covered.

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py:293]: just thinking out loud if it makes sense to keep the build id as well. This would make it easy for us to link to the build directly from these logs, like a link to the build if someone wants to look at the app build docs from here itself.
We can always fetch it from backend using this function `get_build_by_version(domain, app_id, version, return_doc=False)` though.

@orangejenny thoughts? I remember you mentioned you had concerns about storing username in the logs since that can be fetched. This is slightly different here.

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [inline corehq/apps/app_manager/views/releases.py:699]: Can we pass this instead to this function? It seems like it would create this each time for a log.

**Review by mkangia — APPROVED** [review]
No blocking comments, though good to wait for a decision on adding build_id if we agree to before merging.

Requesting @orangejenny  to share her review as well since I have been a part of the design/implementation as well, so the review could be clouded by my own judgement during design.

File: corehq/apps/app_manager/views/releases.py
Comment (pxwxnvermx) [reply corehq/apps/app_manager/views/releases.py:699]: This calculates the time for when each log was created by accessing the `log.created_at` and then adjusting it according to the timezone of user.

https://github.com/dimagi/commcare-hq/blob/7e95cc4e99f67d7f4ea4f39588a0bd251bd608a1/corehq/apps/app_manager/views/releases.py#L685

This function is used here and adds additional data to response.

File: corehq/apps/app_manager/views/releases.py
Comment (orangejenny) [inline corehq/apps/app_manager/views/releases.py:693]: This caught my eye because the calculation looked off (based on my guess of what "more" would be used for) but it looks like it isn't used at all. Should this be removed?

**Review by orangejenny — APPROVED** [review]
Made one more tiny nitpick, but this looks good!

File: corehq/apps/app_manager/views/releases.py
Comment (mkangia) [reply corehq/apps/app_manager/views/releases.py:293]: Ya, definitely no plans of using it off hand. Cool, we can leave this for now

File: corehq/apps/app_manager/tests/views/test_release_build.py
Comment (mkangia) [reply corehq/apps/app_manager/tests/views/test_release_build.py:65]: okay. 

Well even if we verify that the user id is same as the only user, it will still confirm that a user_id is being saved on the record and that too its the logged in user.
We could have the same argument for the app id too, but we are still filtering for it so I think the same intention applies for checking the user.

Checking with another user present, is a good point, though it won't really add any value to the test.

File: corehq/apps/app_manager/tests/views/test_release_build.py
Comment (mkangia) [reply corehq/apps/app_manager/tests/views/test_release_build.py:44]: https://github.com/dimagi/commcare-hq/blob/d416577393329e6ad00e0e4e94fd1376c28355c6/corehq/apps/domain/models.py#L796

confirmed that these are handled as a part of delete function call on Domain object

File: corehq/apps/app_manager/tests/views/test_release_build.py
Comment (pxwxnvermx) [reply corehq/apps/app_manager/tests/views/test_release_build.py:65]: great, i will add the check for user_id, thanks.

**Comment by mkangia** [comment]
:tada:

**Pull request merged**
